android/Staccato_AN/app/src/main/java/com/on/staccato/presentation/staccatoupdate/viewmodel/StaccatoUpdateViewModel.kt

```diff
@@ -31,299 +31,302 @@ import dagger.hilt.android.lifecycle.HiltViewModel
-            }
+    private fun fetchStaccatoBy(staccatoId: Long) {
+        viewModelScope.launch {
+            staccatoRepository.getStaccato(staccatoId = staccatoId)
+                .onSuccess { staccato ->
+                    staccatoTitle.set(staccato.staccatoTitle)
+                    _address.value = staccato.address
+                    _latitude.value = staccato.latitude
+                    _longitude.value = staccato.longitude
+                    _selectedVisitedAt.value = staccato.visitedAt
+                    _placeName.value = staccato.placeName
+                    _currentPhotos.value = createPhotosByUrls(staccato.staccatoImageUrls)
+                    _selectedMemory.value =
+                        MemoryCandidate(
+                            staccato.memoryId,
+                            staccato.memoryTitle,
+                            staccato.startAt,
+                            staccato.endAt,
+                        )
+                }.onException(::handleInitializeException)
+                .onServerError(::handleServerError)
         }
+    }
 
-        private fun createPhotoUploadJob(
-            context: Context,
-            photo: AttachedPhotoUiModel,
-        ) = viewModelScope.async(buildCoroutineExceptionHandler()) {
-            val multiPartBody =
-                convertExcretaFile(
-                    context,
-                    photo.uri,
-                    StaccatoCreationViewModel.FORM_DATA_NAME,
-                )
-            imageRepository.convertImageFileToUrl(multiPartBody)
-                .onSuccess {
-                    updatePhotoWithUrl(photo, it.imageUrl)
-                }.onException(::handleUpdatePhotoException)
+    private fun fetchMemoryCandidates() {
+        viewModelScope.launch {
+            timelineRepository.getMemoryCandidates()
+                .onSuccess { memoryCandidates ->
+                    _memoryCandidates.value = memoryCandidates
+                }.onException(::handleMemoryCandidatesException)
                 .onServerError(::handleServerError)
         }
+    }
 
-        private fun buildCoroutineExceptionHandler(): CoroutineExceptionHandler {
-            return CoroutineExceptionHandler { _, throwable ->
-                _warningMessage.postValue(throwable.message ?: FAIL_IMAGE_UPLOAD_MESSAGE)
+    private fun createPhotoUploadJob(
+        context: Context,
+        photo: AttachedPhotoUiModel,
+    ) = viewModelScope.async(buildCoroutineExceptionHandler()) {
+        val multiPartBody =
+            convertExcretaFile(
+                context,
+                photo.uri,
+                StaccatoCreationViewModel.FORM_DATA_NAME,
+            )
+        imageRepository.convertImageFileToUrl(multiPartBody)
+            .onSuccess {
+                updatePhotoWithUrl(photo, it.imageUrl)
+            }.onException { e, message ->
+                if (this.isActive) handleUpdatePhotoException(e, message)
```

Comment:
찾아보니 다른 블로그에서도 `isActive`를 확인하여 코루틴 작업의 활성화 상태를 판단하고 원하는 작업을 수행하도록 구현하고 있었어요. 저는 괜찮은 처리 방식이라고 생각합니다!

그러다 `ensureActive()`라는 메서드를 알게 되었습니다! 이 메서드를 활용해봐도 좋을 것 같아요.
아래와 같이 `Job`의 확장함수로 구현되어 제공해주고 있습니다.
```kotlin
fun Job.ensureActive(): Unit {
    if (!isActive) {
         throw getCancellationException()
    }
}
```

내부적으로 `isActive`를 확인하여 활성화 상태가 아닌 경우 `JobCancellationException`을 `throw` 하도록 동작합니다. 그래서 `Job` 취소로 인한 `JobCancellationException` 예외가 `coroutineExceptionHandler`에서 잡힐 것으로 예상합니다...!
아래와 같이 활용할 수 있을 것 같습니다!
```suggestion
                ensureActive()
                handleUpdatePhotoException(e, message)
```

아래 레퍼런스를 참고해보시면 좋을 것 같아요!
[https://android-devpia.tistory.com/11](https://android-devpia.tistory.com/11)

Reply:
`JobCancellationException`은 코루틴을 취소하기 위해 사용되는 특별한 예외로, 전파되지 않는다는 특징을 갖고 있다네요!
현재 방식은 isActive를 체크하여 `CancellationException`를 무시하고 있는데 호두가 제안해주신 코드는 이를 무시하지 않고 `handleUpdatePhotoException`에서 한번 더 처리하게 됩니다.
저는 `CancellationException`에 별도의 핸들링이 필요하지 않아 지금 코드로도 충분하다고 생각하는데 호두는 어떻게 생각하시나요?

Reply:
`CancellationException` 이 코루틴에 전파되지 않는다는 사실을 몰랐네요!
저는 `ensureActive()` 내에서 예외를 던지는 `throw getCancellationException()` 으로 인해 코루틴 스코프가 `handleUpdatePhotoException()` 보다 더 먼저 예외를 잡으면서, 네트워크 에러 메시지가 나타나지 않을 것으로 생각했습니다. 😅

> 호두가 제안해주신 코드는 이를 무시하지 않고 handleUpdatePhotoException에서 한번 더 처리하게 됩니다.

그렇다면 지금의 코드가 제일 좋은 구현인 것 같습니다!